### ISSUE TITLE
Make tempfile.* classes path-like

### ISSUE THREAD

**Issue by 8527e2fb-3575-4b29-bda6-022dc83e1251** [python/cpython#87646]
BPO | [43480](https://bugs.python.org/issue43480)
--- | :---
Nosy | @pxeger

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-03-12.15:49:24.927>
labels = ['type-feature', 'library', '3.10']
title = 'Add .path method/property to tempfile.* for a pathlib.Path'
updated_at = <Date 2021-03-12.15:49:24.927>
user = 'https://github.com/pxeger'
```

bugs.python.org fields:
```python
activity = <Date 2021-03-12.15:49:24.927>
actor = 'pxeger'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2021-03-12.15:49:24.927>
creator = 'pxeger'
dependencies = []
files = []
hgrepos = []
issue_num = 43480
keywords = []
message_count = 1.0
messages = ['388540']
nosy_count = 1.0
nosy_names = ['pxeger']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue43480'
versions = ['Python 3.10']
```

</p></details>

### Linked PRs
* gh-104466
* gh-114765

**Comment by barneygale** [comment]
Proposal: make `NamedTemporaryFile` and `TemporaryDirectory` subclasses of `Path`. We'd need to deprecate and remove the existing `name` attribute first as it means something different to pathlib's `name`.

**Comment by AlexWaygood** [comment]
...Is the deprecation cycle for the `name` attribute really worth the hassle, and pain to users? I agree that that's how I might do it if I were writing `tempfile` from scratch today. But the proposed PR has fairly nice semantics, in my opinion, and we could implement it now without having to force any users to modify their code that's currently working today.

While I love it and generally reach for it in my code, it's also the case that not all users will need or want pathlib's interface -- some might be happy with string paths and `os.path`. For these users, reimplementing these classes as `Path` subclasses might lead to an unnecessary performance regression.

**Comment by barneygale** [comment]
I think you're right Alex, and I'm probably being too idealistic. Thanks for the feedback!

**Comment by barneygale** [comment]
I'm still not convinced about this feature request, though. Is this not already easy and explicit?

```python
# currently possible
with NamedTemporaryFile() as tf:
    temppath = Path(tf)

# proposed
with NamedTemporaryFile() as tf:
    temppath = tf.path
```

I don't think the name `.path` makes it as obvious that this is a rich path object. `os.DirEntry` also has a `.path` attribute, but it's a _string_, not a `pathlib.Path` object.

**Comment by AlexWaygood** [comment]
> I'm still not convinced about this feature request, though. Is this not already easy and explicit?
> 
> ```python
> # currently possible
> with NamedTemporaryFile() as tf:
>     temppath = Path(tf)
> 
> # proposed
> with NamedTemporaryFile() as tf:
>     temppath = tf.path
> ```
> 
> I don't think the name `.path` makes it as obvious that this is a rich path object. `os.DirEntry` also has a `.path` attribute, but it's a _string_, not a `pathlib.Path` object.

Yes, I think that's a much better objection. For me, I use `TemporaryDirectory()` a fair bit, and so it's annoying to have to repeat the same boilerplate-y "and now we convert the string to a path immediately" thing over and over again. But that certainly doesn't justify an unclear/confusing API, and it's hard to know what a _good_ API would be here :)

**Comment by barneygale** [comment]
I've just realised that these classes aren't `os.PathLike`. If we add an `__fspath__()` method to them, then this becomes possible:

```python
with NamedTemporaryFile() as tf:
    temppath = Path(tf)
```

I prefer this to a `.path` attribute.

**Comment by barneygale** [comment]
I'm going to adjust the title and remove the `topic-pathlib` tag, because I think there's very little prospect of exposing `pathlib.Path` objects in `tempfile` APIs. We may still wish to make some `tempfile` classes `os.PathLike`.

**Comment by Salamandar** [comment]
I'm struggling with a library that might return either a Path or a TemporaryDict, and having one inheriting the other would really be interesting.

**Comment by serhiy-storchaka** [comment]
I am against this change.

The path-like protocol was specially designed to support pathlib.Path and third-party implementations of path objects (not necessary compatible with pathlib) in functions which need a path argument. Initially it was supposed to use str() to get a string representation of the patch, but str() does not fail for non-paths and does not work as expected with bytes paths, so special protocol was added. You only need to implement it in a class that represents a path.

NamedTemporaryFile and TemporaryDirectory are not paths, they have different purpose. They represent files objects with limited lifespan. You can get their path, but path is not a file and is not a directory. For the same reason file objects, file archive objects, databes objects do not implement the path-like protocol.

In this case it is especially dangerous for these classes because their destructors have serious side effect. Consider the following examples:
```py
def func(paths):
    while paths:
        path = paths.pop()
        with open(path):
            ...
```
```py
def func(path):
    while paths:
        path = paths.pop()
        path = os.fspath(path)
        with open(path):
            ...
```
Passing just created NamedTemporaryFile object to the first function will "work", because it keeps a reference to the NamedTemporaryFile object while working with a file. Passing it to the second function will fail, because the reference to the NamedTemporaryFile object will be lost. The author of that function do not expect a "path" with such side effect. The user of that function did not notice that passing a NamedTemporaryFile object without keeping a reference may be not safe. And there may be more code between.

So please use `__fspath__()` only for objects representing a path.
